docs/reconfigurator.adoc

```diff
@@ -179,6 +179,166 @@ We're being cautious about rolling out that kind of automation.  Instead, today,
+. Add `my_config: Option<MyConfig>` to the blueprint structures (both in-memory and in the database).  This field has to be optional so that when updating to this release, the system can still read the current target blueprint (that was written in the previous release that didn't have this field).
+. In the Reconfigurator planner, when generating a blueprint based on a parent blueprint where `my_config` is `None`, fill in `my_config` (using a `Some` value) based on the contents in inventory.
+. Add `my_config` to the Sled Agent request that will be used by Reconfigurator to _configure_ this on each sled.
+** If a request already exists (e.g., if this will be part of `OmicronZoneConfig` that already gets sent by Reconfigurator to Sled Agent, as in the case of ZFS dataset id or artifact id): the new field should be optional: `my_config: Option<MyConfig>`.  This is required for the system to be able to execute the last target blueprint that was written in the _previous_ release.  This is typically also necessary because it's usually the same struct that Sled Agent records persistently.  See the next item.
+** If no request already exists for this purpose, then you'll be adding a whole new one (e.g., when we added a new `PUT /datasets`).  The body of this request will generally be type `MyConfig` (_not_ optional).  During execution, Reconfigurator can avoid making this request altogether if the blueprint does not specify it.
+. Add `my_config` to the Sled Agent ledger that will store this information persistently.  _This will almost always be the same as the previous step_.  The structure that Sled Agent stores is generally the same one it accepts from Nexus.
++
+This explains another reason why `my_config` should be optional in this structure: Sled Agent _must_ be able to read ledgers written by a previous release and those won't have this field.
+
+**During the upgrade to this release:**
+
+. Wait for at least one inventory cycle to complete successfully and verify that it contains the expected `my_config` field.
+. Generate a new blueprint, make it the current target, and ensure that it executes successfully.  It should make no actual changes to the system, but it will propagate the current values for `my_config` to the blueprint system and to sled ledgers.
+. Verify that:
+** the new blueprint has `my_config` filled in
+** all Sled Agent ledgers have `my_config` filled in (value `Some`)
+
+**In the next release** (we'll call it "release 3"): all the optional fields can be made non-optional:
+
+* Blueprints' in-memory structure can go from `my_config: Option<MyConfig>` to `my_config: MyConfig`.
+* Blueprints' database representation can go from NULL-able columns to non-NULL-able ones, though only if we can populate the value or drop it from old blueprints.  More work is needed here.
```

Comment:
This is another spot where "if we know we added a blueprint during upgrade", it also might give us the leniency to "delete all older blueprints as the final step of an upgrade". Or perhaps as one of the first steps of upgrading to N+1